design-proposals/vm-in-a-container.md

```diff
@@ -0,0 +1,186 @@
+<qemu:commandline>
+   <qemu:env name='kubevirt.io.vm.uuid' value='1234-5678-1234-1234'/>
+   <qemu:env name='kubevirt.io.vm.name' value='testvm'/>
+</qemu:commandline>
+</domain>
+```
+
+Note that we have to specify the qemu namespace to use these features.
+
+Based on this information, the binary can check all runninig processes for the
+matching shim process and derive namespaces and cgroups from it. Then it can
+start the real qemu process inside them.  For namespaces it can use the
+`setns()` system call.
+
+### Security
+
+Since the driver binary will have the `suid` bit set, to do all necessary
+operations. This imposes one big security risk: The binary can start
+applications in other namespaces and cgroups (even the one of the host).
+
+A few things can be done to reduce this risk:
+
+ 1. Only allow the qemu user to execute the binary
+ 2. Introduce a shared secret between the VM target container and kubevirt
+ 3. This secret can be unique per pod creation and only valid for a specific
+    amount of time
+ 4. The binary checks that  the secret is present in the kubernetes metadata
+    file before it starts something in that container 
+
+The secret can be delivered to the container through 
+[Kubernetes secrets](https://kubernetes.io/docs/user-guide/secrets/). The
+binary can receive the secret via libvirts <qemu:commandline/> tag as
+environment variable:
+
+```xml
+<qemu:commandline>
+   <qemu:env name='kubevirt.io.secret' value='3098tFJoswfwkjp4'/>
+</qemu:commandline>
+```
+
+Note that this means that the secret is exposed in the Domain XML. If we think
+that having a secret lifespan of e.g. one minute is not secure enough, the
+secret could be encrypted.
+
+To further improve security, a [libvirt
+hook](https://libvirt.org/hooks.html#qemu) will check if the Domain XML has an
+emulator tag which points to the right binary.
+
+### Flow
+
+Pseudocode:
+
+```
+# Getting the pid os container runtime independent
+pid = $(ps aux  | grep virt-launcher | grep "-kubevirt.vm.uuid 1234-5678-1234-1234 | tr -s ' ' | cut -d " " -f2)
+
+if SecretFromPidMountNamspace(pid) == decrypt(providedSecret) && SecredFromPidMountNamespaceStillValid(pid) {
+  executeQemuInPidNsAndCgroupContext(pid)
+} else {
+  fail()
+}
+```
+
+### Running it inside a container
+
+To run the binary from inside a container, it needs access to the host `/proc`
+directory. The recommended implementaiton is, to have a config file for the
+binary in the container on a well known location (e.g.
+`/etc/kubevirt/qemu-kube.conf`). In there the path to `/proc` can be specified.
+On the specified location, the host `/proc` can be mounted.
+
+### Delivering the binary
+
+At the end, the binary and the libvirt hook need to be present in the mount
+namespace where libvirt is running. They should be delivered via an init
+container to the host via `hostDir` mount (if libvirt is running on the host),
+or via an `emptyDir` mount to a container (if libvirt is running in a
+container).
```

Comment:
Further, virt-handler controlls the additional qemu commandline and environment flags, and therefore how the emulator is called. So for me it is more like a part of virt-handler and that it is more important that it is kept in-sync with virt-handler than with libvirt. What do you think?